loopy/kernel/creation.py

```diff
@@ -1428,9 +1443,9 @@ def add_assignment(base_name, expr, dtype, additional_inames):
 
     for insn, insn_inames_to_dup in zip(instructions, inames_to_dup):
         if isinstance(insn, MultiAssignmentBase):
-            found = [False]
-            new_expression = cseam(insn.expression, frozenset(), found)
-            if found[0]:
+            new_expression = cseam(insn.expression, frozenset())
+            if contains_cse(insn.expression):
```

Comment:
Maybe this condition could just become `if new_expression is not insn.expression`, with no need for another traversal.

Reply:
I don't think IdentityMapper is generous about returning identical expression objects.

Reply:
It should be though. There are 2 options I think,
1. check for identical obejcts as @inducer said
2. instead of passing a found object in the call to the mapper, store in the mapper as an attribute and query it after traversal.

Reply:
> I don't think IdentityMapper is generous about returning identical expression objects.

Turns out I was wrong, it was already quite mature in preserving objects, whenever possible.

Yep, went with option 1.

Thanks for the suggestions!

Reply:
> it was already quite mature in preserving objects whenever possible.

I seem to recall @isuruf did a lot of work towards that goal, so I had a feeling it was worth a shot. Good to hear!